easybuild/easyconfigs/n/NVHPC/NVHPC-20.7-GCC-9.3.0.eb

```diff
@@ -0,0 +1,59 @@
+name = 'NVHPC'
+version = '20.7'
+local_gccver = '9.3.0'
+versionsuffix = '-GCC-%s' % local_gccver
+
+homepage = 'https://developer.nvidia.com/hpc-sdk/'
+description = """C, C++ and Fortran compilers included with the NVIDIA HPC SDK (previously: PGI)"""
+
+toolchain = SYSTEM
+
+# NVHPC can be downloaded freely from NVIDIA's website at https://developer.nvidia.com/hpc-sdk
+# It requires accepting the HPC SDK Software License Agreement (https://docs.nvidia.com/hpc-sdk/eula/index.html)
+# If you accept the License Agreement, you may also uncomment the following line to automatically download the sources
+# source_urls = ['https://developer.download.nvidia.com/hpc-sdk/%(version)s/']
+sources = ['nvhpc_2020_%(version_major)s%(version_minor)s_Linux_x86_64_cuda_multi.tar.gz']
+checksums = ['a5c5c8726d2210f2310a852c6d6e03c9ef8c75e3643e9c94e24909f5e9c2ea7a']
+
+dependencies = [
+    ('GCCcore', local_gccver),
+    ('binutils', '2.34', '', ('GCCcore', local_gccver)),
+    # This is necessary to avoid cases where just libnuma.so.1 is present in the system and -lnuma fails
+    ('numactl', '2.0.13', '', ('GCCcore', local_gccver))
+]
+
+default_cuda_version = None  # Use strings like "11.0" or use the command line: --try-amend=default_cuda_version="10.2"
+
+# NVHPC EasyBlock supports some features, which can be set via CLI or this easyconfig.
+# The following list gives examples for the easyconfig
+#
+# NVHPC needs CUDA to work. Two options are available: 1) Use NVHPC-bundled CUDA, 2) use system CUDA
+# 1) Bundled CUDA
+#    If no easybuild dependency to CUDA is present, the bundled CUDA is taken. A version needs to be specified with
+#      default_cuda_version = "11.0"
+#    in this easyconfig file; alternatively, it can be specified through the command line during installation with
+#      --try-amend=default_cuda_version="10.2"
```

Comment:
@AndiH Having to tweak the `default_cuda_version` value while we use the bundled CUDA is a bit weird...

There's only one CUDA version bundled, right? Shouldn't we set `default_cuda_version` accordingly, and make a note that this has to be tweaked when `CUDA` is included as a dep (setting to `None` would work in that case, since it'll be auto-derived)?

Also, what happens if there's a mismatch?

Reply:
@boegel Nope, if users choose the "nvhpc[…]_Linux_x86_64_cuda_multi.tar.gz" tar ball, it come with the current and two past CUDA versions. There is a tar ball which has only one CUDA bundled, but since that is more specific than this one, I decided to rather focus on the _mutli_ tar ball.

If there is a mismatch, i.e. when the _requested_ CUDA version is not available in the tar ball, it's up to NVHPC's `install.sh` script to error out. We only pass the requested CUDA version to the script via environment variable – everything else is handled by the script.

Reply:
I still have a feeling that some of this is backwards: with the usual HMNS a user would do
`module load NVHPC CUDA`
and then the NVHPC module is loaded first, and the CUDA module *after*, and NVHPC extends MODULEPATH.

But that means that the NVHPC installation procedure has no knowledge which CUDA is eventually used with it! Sure it could default to the builtin but we need to make sure that it's runtime overridable.

Reply:
i.e we need an easyconfig as follows (just to illustrate here what I mean):
```
$ cat CUDA-11.0.2-NVHPC-20.7.eb
easyblock = 'Bundle'
name = 'CUDA'
version = '11.0.2'

homepage = 'https://developer.nvidia.com/cuda-toolkit'
description = """CUDA (formerly Compute Unified Device Architecture) is a parallel
 computing platform and programming model created by NVIDIA and implemented by the
 graphics processing units (GPUs) that they produce. CUDA gives developers access
 to the virtual instruction set and memory of the parallel computational elements in CUDA GPUs."""

toolchain = {'name': 'NVHPC', 'version': '20.7'}

dependencies = [('CUDAcore', '11.0.2', '', True)]

altroot = 'CUDAcore'

moduleclass = 'system'
```

Reply:
The compile-time specified CUDA version goes into a `cudarc` file in the NVHPC directory structure. If an external CUDA is loaded, this is – as far as I know – not taken, but rather deduced by looking at `$CUDA_HOME`. So specifying a different compile-time and run-time CUDA _should_ not be that bad.

I don't have anything against providing a CUDA easyconfig inside the NVHPC toolchain, that would be a good addition for the CUDA-dependency-free NVHPC easyconfig.

For JSC's case, in which we want to more strongly link CUDA and NVHPC, I'd still prefer having a CUDA dependency inside the easyconfig file of NVHPC.

Reply:
I did some experiments with NVHPC 20.7 and looked around a bit: its included CUDA 11.0 is indentical to what the CUDA easyconfig gives us too so there is some overlap here.. on the other hand it seems to ignore CUDA_HOME. I'll dig a bit further.

It's all a little bit of a contrast between NVHPC shipping the world and the way EasyBuild divides full toolchains into components. This is all similar to the Intel cluster toolkit which also includes everything including Intel MPI and MKL, but EB installs those separately. You can argue for both approaches -- it's just more consistent with the rest of EB to reduce NVHPC strictly to the compilers, if at all possible.

Reply:
Exactly, NVIDIA wants to position this as the all-in-one solution for HPC, which is fine, but doesn't really align with the way we currently handle modules in a fine-grained manner.

Reply:
If you use `nvc++ -stdpar` it ignores `$CUDA_HOME` and insists using the included CUDA. This can all be followed in the various rc files (e.g. `$EBROOTNVHPC/20.7/Linux_x86_64/20.7/compilers/bin/rcfiles/paralgorc`) but these files have a scary notice at the top which doesn't invite fixing that issue. I'll get back to you again tomorrow to get a bit more of a clear view.

Reply:
I think there is a good reason for that: the NVHPC-`stdpar` is (/may be) slightly newer than then CUDA `stdpar`. They probably want to make sure that there's no interference.

Reply:
I would like to settle this today, so we can include this easyconfig in the EasyBuild v4.3.1 release which we're pushing out tomorrow.

@AndiH @bartoldeman Any chance you would have time for a quick call to settle this around 4pm CET today (I still need to read up on the discussion)? Or maybe @AndiH can join the EasyBuild Slack (see https://easybuild.readthedocs.io/en/latest/#getting-help) for a more interactive discussion.

Reply:
I've messaged you on Slack.

Reply:
@AndiH I've made the changes we discussed in https://github.com/AndiH/easybuild-easyconfigs/pull/1 .

If you can merge this soon (at the very latest tomorrow morning CET), then we can still squeeze this in for EasyBuild v4.3.1 .
I've changed the target branch for this PR to `4.3.x` to ensure it gets included in the release (cc @migueldiascosta).
Shortly after the release we'll collapse the `4.3.x` branch into `develop` too to sync things up.

Reply:
@boegel's changes are merged!

I'm happy this is shipping with the next EasyBuild release!

Reply:
:+1: